src/libraries/System.Net.NameResolution/src/System/Net/Dns.cs

```diff
@@ -594,19 +594,33 @@ private static Task GetHostEntryOrAddressesCoreAsync(string hostName, bool justR
             static async Task<T> CompleteAsync(Task task, string hostName, long startingTimestamp)
             {
                 _ = NameResolutionTelemetry.Log.BeforeResolution(hostName);
-                T? result = null;
+                string? errorType = null;
                 try
                 {
-                    result = await ((Task<T>)task).ConfigureAwait(false);
-                    return result;
+                    return await ((Task<T>)task).ConfigureAwait(false);
+                }
+                catch (Exception ex)
+                {
+                    errorType = GetErrorType(ex);
+                    throw;
                 }
                 finally
                 {
-                    NameResolutionTelemetry.Log.AfterResolution(hostName, startingTimestamp, successful: result is not null);
+                    NameResolutionTelemetry.Log.AfterResolution(hostName, startingTimestamp, errorType);
                 }
             }
         }
 
+        private static string GetErrorType(Exception exception) => (exception as SocketException)?.SocketErrorCode switch
```

Comment:
These strings are specifically for metrics, right? Please add a comment on the method to that effect.  Where do these result strings come from? Are they strings dictated by otel, or are you just snake_casing the names from the enum?

Reply:
> These strings are specifically for metrics, right?

Yes.

> Where do these result strings come from?

They are defined by us, by mapping enum values to snake-cased strings which is an OTel convention. [We do the same in HTTP metrics](https://github.com/dotnet/runtime/blob/1d760d6de8250eecb5b3711957e0393a796c89b5/src/libraries/System.Net.Http/src/System/Net/Http/Metrics/MetricsHandler.cs#L157-L173).

Reply:
>  which is an OTel convention

Why don't we do the same for the `_ =>` case then? i.e. it seems like our general goal here is to take any SocketError value or Exception name, and snake case it, and we're then just optimizing a few cases that we expect to be most common by hardcoding the transformation. In the name of robustness, shouldn't we be dynamically creating the snak_cased version of the resulting SocketError if there is one (outside of the range we've hardcoded) or for the exception type?

Reply:
And when that mapping is augmented in the future, we're going to remember to update this random method as well?

Reply:
> it seems like our general goal here is to take any SocketError value or Exception name, and snake case it

I'd frame the goal as disambiguating the most common sources of error that users would appreciate not having grouped together when analyzing failures. I don't think it is a goal that every possible SocketError has been disambiguated, that new future SocketError values are automatically disambiguated, or that exception names are transformed into snake case. The hard-coded list Anton has here seems fine to me. If we ever changed it we'd probably want that to be an intentional choice so we can warn users that their queries might need to be updated.